src/database-postgresql.cpp

```diff
@@ -0,0 +1,319 @@
+}
+
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_OK)
+		return;
+
+	PQreset(m_conn);
+	ping();
+}
+
+
+bool Database_PostgreSQL::ping()
+{
+	if (PQping(m_conninfo.c_str()) != PQPING_OK) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+	return true;
+}
+
+
+bool Database_PostgreSQL::initialized()
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+
+void Database_PostgreSQL::prepare()
+{
+	if (m_pgVersion < 90500) {
+		prepareStatement("save",
+			"INSERT INTO blocks (x, y, z, data) VALUES ($1::int, "
+			"$2::int, $3::int, $4::bytea)");
+		prepareStatement("update",
+			"UPDATE blocks SET data = $4::bytea WHERE x = $1::int "
+			"AND y = $2::int AND z = $3::int");
+	} else {
+		prepareStatement("save", "INSERT INTO blocks (x, y, z, data) "
+			"VALUES($1::int, $2::int, $3::int, $4::bytea) ON "
+			"CONFLICT UPDATE SET data = $4::bytea;");
+	}
+	prepareStatement("load",
+		"SELECT data FROM blocks WHERE x = $1::int AND y = $2::int AND "
+		"z = $3::int");
+	prepareStatement("delete",
+		"DELETE FROM blocks WHERE x = $1::int AND y = $2::int AND "
+		"z = $3::int");
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+
+void Database_PostgreSQL::prepareStatement(std::string name, std::string sql)
+{
+	resultsCheck(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+}
+
+
+bool Database_PostgreSQL::resultsCheck(PGresult* res, bool clear /*= true*/,
+	ExecStatusType expected /*= PGRES_COMMAND_OK*/)
+{
+	if (PQresultStatus(res) != expected) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ")
+			+ (res ? PQresultErrorMessage(res) :
+			"Result is NULL!"));
+	}
+
+	if (clear)
+		PQclear(res);
+	else
+		m_res = res;
+
+	return true;
+}
+
+
+bool Database_PostgreSQL::execPrepared(std::string name, int nParams,
+	const v3s16 &pos, bool clear /*= true*/,
+	const std::string &data /*= ""*/,
+	ExecStatusType expected /*= PGRES_COMMAND_OK*/)
+{
+	const char *values[nParams];
+	int         lengths[nParams];
+	int         formats[nParams];
+
+	values[0] = itos(pos.X).c_str();
+	values[1] = itos(pos.Y).c_str();
+	values[2] = itos(pos.Z).c_str();
+
+	lengths[0] = sizeof(values[0]);
+	lengths[1] = sizeof(values[1]);
+	lengths[2] = sizeof(values[2]);
+
+	formats[0] = 0;
+	formats[1] = 0;
+	formats[2] = 0;
+
+	if (nParams == 4) {
+		values[3]  = data.c_str();
+		lengths[3] = data.length();
+		formats[3] = 1;
+	}
+
+	return resultsCheck(PQexecPrepared(m_conn, name.c_str(), nParams,
+		values, lengths, formats, 1), clear, expected);
+}
+
+
+void Database_PostgreSQL::createDatabase()
+{
+	resultsCheck(PQexec(m_conn,
+		"SELECT relname FROM pg_class WHERE relname='blocks';"), false,
+		PGRES_TUPLES_OK);
+
+	if (PQntuples(m_res)) {
+		PQclear(m_res);
+		return;
+	}
+
+	resultsCheck(PQexec(m_conn,
+		"SELECT set_config('client_min_messages', 'error', true);\n"
```

Comment:
I do not think it is needed any longer after the check. My paranoia must have been set to extreme.